agent/consul/state/intention.go

```diff
@@ -990,7 +990,13 @@ func (s *Store) intentionTopologyTxn(tx ReadTxn, ws memdb.WatchSet,
 	//				 Ideally those should be excluded as well, since they can't be upstreams/downstreams without a proxy.
 	//				 Maybe narrow serviceNamesOfKindTxn to services represented by proxies? (ingress, sidecar-
 	wildcardMeta := structs.WildcardEnterpriseMetaInPartition(structs.WildcardSpecifier)
-	index, services, err := serviceNamesOfKindTxn(tx, ws, structs.ServiceKindTypical, *wildcardMeta)
+	var services []*KindServiceName
+	if intentionTarget == structs.IntentionTargetService {
+		index, services, err = serviceNamesOfKindTxn(tx, ws, structs.ServiceKindTypical, *wildcardMeta)
+	} else {
+		// destinations can only ever be upstream, since they are only allowed as intention destination.
+		index, services, err = serviceNamesOfKindTxn(tx, ws, structs.ServiceKindDestination, *wildcardMeta)
+	}
```

Comment:
General Q: Why if-else now? Compared to before where they were appended

Reply:
In the other PR (that @acpana is reviewing) I was intending to merge both in the same RPC endpoint and return a list of destinations and services, but I found out that this would be complex as it return a list of serviceNames which will make it hard to differentiate between the two later in the ProxyCfg.
So I chose to go with a separate RPC endpoint to avoid that and also avoid the retro compatibility dilemma.